src/mesh/mesh_dual_weighted_residual_error.cpp

```diff
@@ -0,0 +1,388 @@
+{
+    // assuming that all pointers are still valid
+    // reinitilizing all variables after triangulation in the constructor
+    solution_coarse = dg->solution;
+    adjoint_state = AdjointStateEnum::coarse;
+
+    // storing the original FE degree distribution
+    coarse_fe_index.reinit(dg->triangulation->n_active_cells());
+    
+    // looping over the cells
+    for(auto cell = dg->dof_handler.begin_active(); cell != dg->dof_handler.end(); ++cell)
+        if(cell->is_locally_owned())
+            coarse_fe_index[cell->active_cell_index()] = cell->active_fe_index();
+
+    // for remaining, clearing the values
+    dIdw_fine      = dealii::LinearAlgebra::distributed::Vector<real>();
+    dIdw_coarse    = dealii::LinearAlgebra::distributed::Vector<real>();
+    adjoint_fine   = dealii::LinearAlgebra::distributed::Vector<real>();
+    adjoint_coarse = dealii::LinearAlgebra::distributed::Vector<real>();
+
+    dual_weighted_residual_fine = dealii::Vector<real>();
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+void DualWeightedResidualError<dim, nstate, real, MeshType>::convert_to_state(AdjointStateEnum state, std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{   
+    // checks if conversion is needed
+    if(adjoint_state == state) 
+        return;
+
+    // then calls corresponding function for state conversions
+    if(adjoint_state == AdjointStateEnum::coarse && state == AdjointStateEnum::fine) 
+        coarse_to_fine(dg);
+    
+    if(adjoint_state == AdjointStateEnum::fine && state == AdjointStateEnum::coarse)
+        fine_to_coarse(dg);
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+void DualWeightedResidualError<dim, nstate, real, MeshType>::coarse_to_fine(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{
+    dealii::IndexSet locally_owned_dofs, locally_relevant_dofs;
+    locally_owned_dofs =  dg->dof_handler.locally_owned_dofs();
+    dealii::DoFTools::extract_locally_relevant_dofs(dg->dof_handler, locally_relevant_dofs);
+
+    // dealii::LinearAlgebra::distributed::Vector<double> solution_coarse(dg->solution);
+    solution_coarse.update_ghost_values();
+    
+    // Solution Transfer to fine grid
+    using VectorType       = typename dealii::LinearAlgebra::distributed::Vector<double>;
+    using DoFHandlerType   = typename dealii::DoFHandler<dim>;
+    using SolutionTransfer = typename MeshTypeHelper<MeshType>::template SolutionTransfer<dim,VectorType,DoFHandlerType>;
+
+    SolutionTransfer solution_transfer(dg->dof_handler);
+    solution_transfer.prepare_for_coarsening_and_refinement(solution_coarse);
+
+    dg->high_order_grid->prepare_for_coarsening_and_refinement();
+    dg->triangulation->prepare_coarsening_and_refinement();
+
+    for (auto cell = dg->dof_handler.begin_active(); cell != dg->dof_handler.end(); ++cell)
+        if (cell->is_locally_owned()) 
+            cell->set_future_fe_index(cell->active_fe_index()+1);
```

Comment:
You probably want to check that DG is able to handle that +1 polynomial degree. This is just waiting for a future crash.

Also, curly brackets.

Reply:
Yeah, added an assert statement at the beginning of this coarse_to_fine() in the new commit.

Reply:
1. This will not actually catch it in Release mode. 
2. You're increasing the fe_index *after* checking if it's too big. 
 
This is the type of stuff that a unit test should catch. Right now we only have integration tests.

Reply:
Hmm yeah, assert doesn't seem to work. 

I have replaced it with std::abort() in the new commit and that works well.